src/minimization.cpp

```diff
@@ -0,0 +1,211 @@
+      }
+      for (int k = 0; k < bSucc.size(); ++k) {
+        node->succ[k] = bSucc[k].second;
+      }
+    }
+  }
+}
+
+void barycentricValueDefinition(Net &net, TreeNode *node, int increment) {
+  int connectionsToAdjacentLayer = 0;
+  float rank = 0;
+  for (int k = 0; k < node->pred.size(); ++k) {
+    if (net.getNode(node->pred[k])->layer + increment != node->layer)
+      continue;
+    connectionsToAdjacentLayer += 1;
+    int index = net.getNode(node->pred[k])->number;
+    int portIndex = getPortIndex(node->id, * net.getNode(node->pred[k]));
+    if (increment == 1) {
+      rank += forwardRankDefinition( *net.getNode(node->pred[k]), index, portIndex);
+    } else {
+      rank += backwardRankDefinition( *net.getNode(node->pred[k]), index, portIndex);
+    }
+  }
+  for (int k = 0; k < node->succ.size(); ++k) {
+    if (net.getNode(node->succ[k])->layer + increment != node->layer)
+      continue;
+    connectionsToAdjacentLayer += 1;
+    int index = net.getNode(node->succ[k])->number;
+    int portIndex = getPortIndex(node->id, *net.getNode(node->succ[k]));
+    if (increment == 1) {
+      rank += forwardRankDefinition( *net.getNode(node->pred[k]), index, portIndex);
+    } else {
+      rank += backwardRankDefinition( *net.getNode(node->succ[k]), index, portIndex);
+    }
+  }
+  node->barycentricValue = rank / connectionsToAdjacentLayer;
+}
+
+void layerSweepAlgorithm(Net &net) {
+  std::vector<std::vector<TreeNode::nodeId>> nodesByLayer = net.getNodesByLayer();
+  std::vector<std::vector<TreeNode::nodeId>> tempNodesByLayer(nodesByLayer.size());
+  std::copy(nodesByLayer.begin(), nodesByLayer.end(), tempNodesByLayer.begin());
+
+  int intersections = -1;
+  while (true) {
+    //            forward layer sweeps
+    for (int i = 1; i < tempNodesByLayer.size(); ++i) {
+      for (int j = 0; j < tempNodesByLayer[i].size(); ++j) {
+        barycentricValueDefinition(net, net.getNode(tempNodesByLayer[i][j]), 1);
+      }
+      sortNodes(net, tempNodesByLayer[i]);
+    }
+
+    int intersectionsAfterFls = crossCounting(net, tempNodesByLayer);
+
+    if (intersections > intersectionsAfterFls || intersections == -1) {
+      intersections = intersectionsAfterFls;
+      std::copy(tempNodesByLayer.begin(), tempNodesByLayer.end(), nodesByLayer.begin());
+    } else {
+      break;
+    }
+    //            backwards layer sweeps
+    for (int i = tempNodesByLayer.size() - 2; i >= 0; i--) {
+      for (int j = 0; j < tempNodesByLayer[i].size(); ++j) {
+        barycentricValueDefinition(net, net.getNode(tempNodesByLayer[i][j]), -1);
+      }
+        sortNodes(net, tempNodesByLayer[i]);
+    }
+    int intersectionsAfterBls = crossCounting(net, tempNodesByLayer);
+
+    if (intersections > intersectionsAfterBls || intersections == -1) {
+      intersections = intersectionsAfterBls;
+      std::copy(tempNodesByLayer.begin(), tempNodesByLayer.end(), nodesByLayer.begin());
+    } else {
+        break;
+    }
```

Comment:
I understand the loops are a bit different, but intermediate calculations are the same, i.e. code duplication.